container/ClearSky/src/main/java/io/egen/clearskyboot/services/implementations/WeatherServiceImpl.java

```diff
@@ -0,0 +1,175 @@
+	}
+
+	@Override
+	public List<String> findDistinctCities() {
+		return weatherRepository.findDistinctCities();
+	}
+
+	@Override
+	public Reading findLatestWeatherByCity(String city) {
+		Optional<Reading> optionalReading = Optional.ofNullable(weatherRepository.findTop1ByCityOrderByTimestampDesc(city));
+		if(optionalReading.isPresent()) return optionalReading.get();
+		throw new NotFoundException("No latest data found for city : " + city);
+	}
+
+	@Override
+	public JsonNode findLatestWeatherPropertyByCity(String city, String property) {
+		
+		// Create an object mapper, object node and put city name in the node
+		ObjectMapper objectMapper = new ObjectMapper();
+		
+		Optional<Reading> optionalLatestReading = Optional.ofNullable(weatherRepository.findTop1ByCityOrderByTimestampDesc(city));
+		if(optionalLatestReading.isPresent()) {
+			
+			Reading latestReading = optionalLatestReading.get();
+			boolean propertyNotFound = true;
+			BeanInfo beanInfo;
+			
+			Optional<BeanInfo> optionalBeanInfo;
+			try {
+				optionalBeanInfo = Optional.of(Introspector.getBeanInfo(Reading.class));
+				if(optionalBeanInfo.isPresent()) beanInfo = optionalBeanInfo.get();
+				else throw new NotFoundException("BeanInfo instance not found...");
+			} catch (IntrospectionException e1) {
+				throw new InternalServerError(e1.getMessage());
+			}
+			
+			// Get attributes of class using getBeanInfo
+			// Iterate through attributes to find specific property
+			// Ex. temperature
+			for (PropertyDescriptor propertyDesc : beanInfo.getPropertyDescriptors()) {
+				
+				// If found, add property to objectNode and return objectNode (as JsonNode)
+			    if(propertyDesc.getName().equalsIgnoreCase(property)) {
+			    	
+			    	propertyNotFound = false;
+			    	
+					ObjectNode resultNode = objectMapper.createObjectNode();
+					resultNode.put("city", city);
+					
+					Optional<Method> optionalMethod = Optional.of(propertyDesc.getReadMethod());
+					optionalMethod.ifPresent(m -> {
+						Optional<String> optionalPropertyValue;
+						try {
+							optionalPropertyValue = Optional.of(String.valueOf(m.invoke(latestReading)));
+							if (optionalPropertyValue.isPresent()) resultNode.put(property, optionalPropertyValue.get());
+					    	else resultNode.put(property, "Value not found");
+						} catch (Exception e) {
+							throw new NotFoundException("Value for property: " + property + ", not found");
+						}
+					});
+					
+					return resultNode;
+			    } // End of IF
+			} // End of FOR
+			
+			// If the property we're searching is an invalid one, throw NotFoundException
+			if(propertyNotFound) throw new NotFoundException("Property " + property + " does not exist!");
+		} else throw new NotFoundException("No data found for city: " + city);
+		return objectMapper.createObjectNode(); // Empty Node
+	}
+	
+	@Override
+	public JsonNode findAvgWeatherPropertyByCityAndGrain(String city, String property, String grain) {
```

Comment:
Can we return a well defined POJO instead of a JsonNode. A POJO will help immensely if this service is used internally elsewhere.